microcosm_flask/basic_auth.py

```diff
@@ -0,0 +1,80 @@
+"""
+Basic Auth support.
+
+Usage:
+
+    @graph.app.route("/path")
+    @graph.basic_auth.required
+    def must_be_authorized():
+        pass
+
+
+"""
+from base64 import b64encode
+
+from flask.ext.basicauth import BasicAuth
+from werkzeug.exceptions import Unauthorized
+
+from microcosm.api import defaults
+from microcosm_flask.errors import with_headers
+
+
+def encode_basic_auth(username, password):
+    """
+    Encode basic auth credentials.
+
+    """
+    return "Basic {}".format(
+        b64encode("{}:{}".format(
+            username,
+            password,
+        ))
+    )
+
+
+class ConfigBasicAuth(BasicAuth):
+    """
+    Basic auth decorator that pulls credentials from static configuration.
+
+    This decorator is sufficient for internal service access control, but should
+    not be used for anything truly sensitive.
+
+    """
+
+    def __init__(self, app, credentials):
+        super(ConfigBasicAuth, self).__init__(app)
+        self.credentials = credentials
+
+    def check_credentials(self, username, password):
+        """
+        Override credential checking to use configured credentials.
+
+        """
+        return password is not None and self.credentials.get(username, None) == password
+
+    def challenge(self):
+        """
+        Override challenge to raise an exception that will trigger regular error handling.
+
+        """
+        response = super(ConfigBasicAuth, self).challenge()
```

Comment:
The default behavior for this library was to return a Flask `Response` with a 401 status code. That works, but won't exercise our error handler logic and return JSON responses.

I'm raising my own error and injecting the expected `WWW-Authenticate` header with this `with_headers` wrapper (and accompanying changes in the error handling).